File: apps/mobile-native/android/app/src/main/java/mx/nexara/mobile/nativeapp/ui/integra/schedules/ScheduleWeekView.kt
package mx.nexara.mobile.nativeapp.ui.integra.schedules

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import mx.nexara.mobile.nativeapp.data.integra.schedules.AcsTime
import mx.nexara.mobile.nativeapp.data.integra.schedules.ScheduleDayPlan
import mx.nexara.mobile.nativeapp.data.integra.schedules.ScheduleTemplate
import mx.nexara.mobile.nativeapp.data.integra.schedules.WeekDay
import mx.nexara.mobile.nativeapp.ui.enterprise.NxColors

/**
 * Vista semanal de una plantilla ACS: siete barras de 24 h con las franjas
 * reales que publica el `WeekPlanCfg` del terminal.
 *
 * Las horas se pintan **tal cual las manda el ACS**, sin pasar por ninguna zona
 * horaria: son el reloj de pared del terminal instalado en sitio. Convertirlas
 * a la zona del teléfono movería la franja de oficina a las 02:00.
 *
 * Cuando el terminal no publica el detalle se dice explícitamente, en lugar de
 * dibujar un horario por convención: una barra inventada que resulta ser falsa
 * es peor que un hueco honesto.
 */
@Composable
fun ScheduleWeekGrid(
    template: ScheduleTemplate?,
    modifier: Modifier = Modifier,
) {
    if (template == null || template.id == ScheduleTemplate.NO_ACCESS_ID) {
        Text(
            "Sin acceso en esta puerta — no abre en ningún horario.",
            style = MaterialTheme.typography.bodySmall,
            color = NxColors.Muted,
            modifier = modifier.padding(vertical = 6.dp),
        )
        return
    }
    if (template.days.isEmpty()) {
        Column(modifier.padding(vertical = 6.dp), verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Text(
                "El terminal no publica el detalle semanal de «${template.name}».",
                style = MaterialTheme.typography.bodySmall,
                color = NxColors.Muted,
            )
            template.weekPlanNo?.let {
                Text(
                    "Plantilla ${template.id} → plan semanal $it",
                    style = MaterialTheme.typography.labelSmall,
                    color = NxColors.Muted,
                )
            }
        }
        return
    }

    Column(modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(6.dp)) {
        Text(
            "Horario semanal · ${template.name}",
            style = MaterialTheme.typography.labelMedium.copy(fontWeight = FontWeight.SemiBold),
            color = NxColors.Slate,
        )
        WeekDay.entries.forEach { day ->
            val segments = template.days.firstOrNull { it.week == day }
            DayRow(day = day, plan = segments)
        }
        Text(
            "Franjas leídas del terminal (hora del ACS en sitio, sin convertir).",
            style = MaterialTheme.typography.labelSmall,
            color = NxColors.Muted,
        )
    }
}

@Composable
private fun DayRow(day: WeekDay, plan: ScheduleDayPlan?) {
    val segments = plan?.segments.orEmpty()
    val text = if (segments.isEmpty()) {
        "Cerrado"
    } else {
        segments.joinToString(" · ") {
            "${AcsTime.acsHhMm(it.beginTime)}–${AcsTime.acsHhMm(it.endTime)}"
        }
    }
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .semantics { contentDescription = "${day.label}: $text" },
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            day.short,
            style = MaterialTheme.typography.labelSmall,
            color = NxColors.Muted,
            modifier = Modifier.width(30.dp),
        )
        BoxWithConstraints(
            modifier = Modifier
                .weight(1f)
                .height(14.dp)
                .background(NxColors.Surface, RoundedCornerShape(4.dp)),
        ) {
            val trackWidth = maxWidth
            segments.forEach { segment ->
                val geometry = AcsTime.bandGeometry(segment.beginTime, segment.endTime)
                if (geometry != null) {
                    val (start, width) = geometry
                    Box(
                        modifier = Modifier
                            .offset(x = trackWidth * start)
                            .width(trackWidth * width)
                            .height(14.dp)
                            .background(NxColors.Teal, RoundedCornerShape(4.dp)),
                    )
                }
            }
        }
        Text(
            text,
            style = MaterialTheme.typography.labelSmall,
            color = if (segments.isEmpty()) NxColors.Muted else NxColors.Slate,
            modifier = Modifier.width(112.dp),
        )
    }
}

/** Resumen de una línea; `null` cuando el ACS no publicó franjas. */
@Composable
fun ScheduleWeekSummary(template: ScheduleTemplate?, modifier: Modifier = Modifier) {
    val summary = template?.summary
    Text(
        text = summary ?: "Sin franjas publicadas por el terminal",
        style = MaterialTheme.typography.labelSmall,
        color = NxColors.Muted,
        modifier = modifier,
    )
}

File: apps/mobile-native/ios/NexaraApp/UI/Integra/ScheduleWeekView.swift
import SwiftUI

/// Día ACS tal cual lo publica `WeekPlanCfg` (claves en inglés).
enum ScheduleWeekDay: String, CaseIterable, Identifiable {
    case monday = "Monday"
    case tuesday = "Tuesday"
    case wednesday = "Wednesday"
    case thursday = "Thursday"
    case friday = "Friday"
    case saturday = "Saturday"
    case sunday = "Sunday"

    var id: String { rawValue }
    var short: String {
        switch self {
        case .monday: return "Lun"
        case .tuesday: return "Mar"
        case .wednesday: return "Mié"
        case .thursday: return "Jue"
        case .friday: return "Vie"
        case .saturday: return "Sáb"
        case .sunday: return "Dom"
        }
    }

    static func fromKey(_ raw: String?) -> ScheduleWeekDay? {
        guard let v = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !v.isEmpty else {
            return nil
        }
        return Self.allCases.first { $0.rawValue.caseInsensitiveCompare(v) == .orderedSame }
    }
}

struct ScheduleSegment: Hashable {
    let beginTime: String
    let endTime: String
}

struct ScheduleDayPlan: Hashable {
    let week: ScheduleWeekDay
    let segments: [ScheduleSegment]
}

/// Utilidades de hora de pared ACS (sin conversión de zona).
enum AcsTimeUI {
    static func acsHhMm(_ raw: String?) -> String {
        let value = raw?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if value.isEmpty { return "??:??" }
        if let r = value.range(of: #"(\d{1,2}):(\d{2})"#, options: .regularExpression) {
            let parts = value[r].split(separator: ":")
            guard parts.count >= 2,
                  let h = Int(parts[0]), let m = Int(parts[1]) else { return value }
            return String(format: "%02d:%02d", h, m)
        }
        return value
    }

    static func minutesOfDay(_ raw: String?) -> Int? {
        let value = raw?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard let r = value.range(of: #"(\d{1,2}):(\d{2})"#, options: .regularExpression) else {
            return nil
        }
        let parts = value[r].split(separator: ":")
        guard parts.count >= 2, let h = Int(parts[0]), let m = Int(parts[1]),
              h >= 0, m >= 0, m <= 59 else { return nil }
        return min(1440, h * 60 + m)
    }

    /// Franja como fracción del día `[inicio, ancho]` en 0…1. Fin `00:00` = medianoche siguiente.
    static func bandGeometry(beginTime: String?, endTime: String?) -> (start: CGFloat, width: CGFloat)? {
        guard let a = minutesOfDay(beginTime), let rawEnd = minutesOfDay(endTime) else { return nil }
        let b = rawEnd <= a ? 1440 : rawEnd
        let start = CGFloat(a) / 1440
        let width = max(0.01, CGFloat(b - a) / 1440)
        return (start, width)
    }
}

/// Vista semanal de una plantilla ACS: siete barras de 24 h con franjas reales.
struct ScheduleWeekGrid: View {
    let templateId: String
    let templateName: String
    let weekPlanNo: Int?
    let days: [ScheduleDayPlan]

    var body: some View {
        if templateId == "0" {
            Text("Sin acceso en esta puerta — no abre en ningún horario.")
                .font(.caption)
                .foregroundStyle(.secondary)
                .padding(.vertical, 6)
        } else if days.isEmpty {
            VStack(alignment: .leading, spacing: 4) {
                Text("El terminal no publica el detalle semanal de «\(templateName)».")
                    .font(.caption)
                    .foregroundStyle(.secondary)
                if let weekPlanNo {
                    Text("Plantilla \(templateId) → plan semanal \(weekPlanNo)")
                        .font(.caption2)
                        .foregroundStyle(.secondary)
                }
            }
            .padding(.vertical, 6)
        } else {
            VStack(alignment: .leading, spacing: 6) {
                Text("Horario semanal · \(templateName)")
                    .font(.subheadline.weight(.semibold))
                ForEach(ScheduleWeekDay.allCases) { day in
                    let plan = days.first { $0.week == day }
                    ScheduleDayRow(day: day, plan: plan)
                }
                Text("Franjas leídas del terminal (hora del ACS en sitio, sin convertir).")
                    .font(.caption2)
                    .foregroundStyle(.secondary)
            }
        }
    }
}

private struct ScheduleDayRow: View {
    let day: ScheduleWeekDay
    let plan: ScheduleDayPlan?

    private var segments: [ScheduleSegment] { plan?.segments ?? [] }

    private var text: String {
        if segments.isEmpty { return "Cerrado" }
        return segments.map {
            "\(AcsTimeUI.acsHhMm($0.beginTime))–\(AcsTimeUI.acsHhMm($0.endTime))"
        }.joined(separator: " · ")
    }

    var body: some View {
        HStack(spacing: 8) {
            Text(day.short)
                .font(.caption2)
                .foregroundStyle(.secondary)
                .frame(width: 28, alignment: .leading)
            GeometryReader { geo in
                ZStack(alignment: .leading) {
                    RoundedRectangle(cornerRadius: 4)
                        .fill(Color(.tertiarySystemFill))
                    ForEach(Array(segments.enumerated()), id: \.offset) { _, segment in
                        if let g = AcsTimeUI.bandGeometry(
                            beginTime: segment.beginTime,
                            endTime: segment.endTime
                        ) {
                            RoundedRectangle(cornerRadius: 4)
                                .fill(Color.teal)
                                .frame(width: geo.size.width * g.width)
                                .offset(x: geo.size.width * g.start)
                        }
                    }
                }
            }
            .frame(height: 14)
            Text(text)
                .font(.caption2)
                .foregroundStyle(segments.isEmpty ? .secondary : .primary)
                .frame(width: 110, alignment: .leading)
                .lineLimit(2)
        }
        .accessibilityLabel("\(day.short): \(text)")
    }
}

/// Extrae franjas por weekPlan id desde `devices[].weekPlans` del catálogo.
enum ScheduleWeekCatalog {
    static func daysByWeekPlan(from catalog: [String: Any]) -> [Int: [ScheduleDayPlan]] {
        var out: [Int: [ScheduleDayPlan]] = [:]
        let devices = IntegraJSON.asMapList(catalog["devices"]) ?? []
        for device in devices {
            for plan in IntegraJSON.asMapList(device["weekPlans"]) ?? [] {
                guard let planId = plan.integraInt("id"), out[planId] == nil else { continue }
                let days = daysFromSegments(IntegraJSON.asMapList(plan["enabledSegments"]) ?? [])
                if !days.isEmpty { out[planId] = days }
            }
        }
        return out
    }

    static func weekPlanNoByTemplate(from catalog: [String: Any]) -> [String: Int] {
        var out: [String: Int] = [:]
        let devices = IntegraJSON.asMapList(catalog["devices"]) ?? []
        for device in devices {
            for tpl in IntegraJSON.asMapList(device["templates"]) ?? [] {
                guard let id = tpl.integraStr("id", "planTemplateNo") else { continue }
                if out[id] == nil, let no = tpl.integraInt("weekPlanNo") {
                    out[id] = no
                }
            }
        }
        return out
    }

    static func daysFromSegments(_ segments: [[String: Any]]) -> [ScheduleDayPlan] {
        var byDay: [ScheduleWeekDay: [ScheduleSegment]] = [:]
        for row in segments {
            guard let day = ScheduleWeekDay.fromKey(row.integraStr("week")),
                  let begin = row.integraStr("beginTime"),
                  let end = row.integraStr("endTime") else { continue }
            byDay[day, default: []].append(ScheduleSegment(beginTime: begin, endTime: end))
        }
        return ScheduleWeekDay.allCases.compactMap { day in
            guard let segs = byDay[day], !segs.isEmpty else { return nil }
            return ScheduleDayPlan(
                week: day,
                segments: segs.sorted { $0.beginTime < $1.beginTime }
            )
        }
    }
}
